### DISCUSSION TITLE
WSL startup directory (WT vs. Start\Run)

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#13877] [General]
In WT, my "Ubuntu" profile has command line "wsl.exe" and no startup directory specified.  When I start it, I'm in

`/home/vefatica`

When I use the Start\Run dialog to start wsl.exe, I'm in

`/mnt/c/Users/vefatica`

What accounts for the difference?

Note that WT itself has CWD v:\ (a SUBST for d:\work).

Thanks.

**Comment by zadjii-msft** [comment]
That's a good question. What's the raw json for the profile? 

I'm pretty sure we default WSL distro `startingDirectory`s to `~`, so if none is set, then we'll treat that as `~`. We do that by manually inserting `~` into the `wsl` commanline.

Just starting `wsl.exe` from the run dialog or commandline wouldn't do the same thing, however. It'll just start in the CWD (which IIRC the run dialog uses `%userprofile%` as the CWD)

**Comment by vefatica** [comment]
I think I get it.  And it probably has nothing to do with WT.

More experimenting suggests that the "Ubuntu" profile (with no startup directory specified) inherits WT's CWD.  I can change WT's CWD on the fly and Ubuntu's startup directory follows (but not when it's a SUBST'd drive).  When WT's CWD is a SUBST'd drive I reckon I'm getting my WSL shell's default startup directory.  

Here's my profile.

```
"commandline": "wsl.exe",
"guid": "{4ba2fd44-db0a-44c1-95c2-ee4e55916fa1}",
"hidden": false,
"name": "**Ubuntu**"

```

**Comment by vefatica** [comment]
P.S., I noted long ago that WSL automatically gets Windows drives in /mnt ... but not SUBST'd drives.  (Edit) I just found discussions of that in a few places.

**Comment by vefatica** [comment]
Hmmm!  If d:\gnu is WT's CWD and I click "**Ubuntu**" in the dropdown menu (you have seen the JSON), I see this (same in TaskManager).  It looks pretty deliberate.

![image](https://user-images.githubusercontent.com/61856645/187342668-29529b69-f9d0-4a56-a9e5-c9b83687bdb9.png)

That test was with 

Microsoft Windows 10 Pro for Workstations
10.0.19044.1889 (2009, 21H2)
WindowsTerminalPreview 1.15.2208.16002

**Reply by DHowett** [reply]
We intentionally move the starting directory value -- whether it came in from `wt.exe` or was set in the profile -- into an argument for WSL. This is so that you can set things like `startingDirectory: /home/vince` and have it work the way you might expect.

As you might imagine, if we _did not_ do that... `CreateProcess()` would fail because `/home/vince` does not exist in the Windows filesystem namespace.

I can't comment on WSL ignoring `SUBST`ituted drives, as I have no particular experience with them.

**Comment by vefatica** [comment]
It's using WT's current directory; that just doesn't make sense.  It might be anything (and quite unrelated to WSL).  Using "~" (as Mike mentioned) when no startup directory is specified, makes sense.  Using "~" is even mentioned in WSL.EXE's command line help.

**Reply by DHowett** [reply]
If you have created _your own WSL profile_ (and you are not using the one detected by Terminal; I believe this to be the case because the GUID you specified does not match any of the WSL detector's GUIDs), you're hitting the default starting directory rules.

The default starting directory for any profile that does not (1) inherit it from somewhere else or (2) set one explicitly is: `null`. When it is `null`, it means "Terminal's CWD".	

I would guess that your profile does not (1) inherit a starting directory from any other profile or (2) explicitly set one.

**Comment by vefatica** [comment]
I see.  If I don't specify a startup directory, I get

![image](https://user-images.githubusercontent.com/61856645/187498107-7c7dbe8b-edd7-44ca-83bd-a0919745ee53.png)

Does using WT's CWD help with any "WT here" (context menu) mechanism?

**Reply by DHowett** [reply]
It doesn't (the shell extension and/or third party registry customizations typically pass an explicit starting directory using `-d`). The scenarios it helps with are:

1. The run dialog
2. Launching `wt` by entering it in the Explorer address bar
3. Launching `wt` from an existing shell session, where it can inherit the CWD of the shell.

**Comment by vefatica** [comment]
Then it seems odd that "~" is not the default, with user-specified startup directory and following WT being optional alternatives.

**Reply by DHowett** [reply]
(please try to reply in the chain using the "Write a reply" button; doing otherwise makes it difficult to follow these discussion threads!)

The default _is_ `~` unless you create the profile yourself. Then there is no default.

**Reply by DHowett** [reply]
It is expected that people producing profiles observe the options available to them :smile:

**Reply by vefatica** [reply]
> It is expected that people producing profiles observe the options available to them

There's something to be said for well-chosen defaults.

I was quite accustomed to Ubuntu starting in /home/vefatica.  Knowing I hadn't specified that directory, I figured it was the default (a well-chosen default).  Then I discovered I only got /home/vefatica because WSL didn't understand the SUBST'd v:\ which is almost always WT's startup directory.  Hence this discussion.  It's no big deal.

**Reply by DHowett** [reply]
> There's something to be said for well-chosen defaults.

You're not wrong about that :)
